File: ios/APEXNative/APEX/Core/Engine/WorkoutInsights.swift
import Foundation

enum WorkoutInsights {
    struct Summary: Equatable, Sendable {
        let from: String
        let to: String
        let workouts: Int
        let activeDays: Int
        let durationMinutes: Int
        let activeEnergyKcal: Double?
        let sets: Int
        let reps: Int
        let volumeKG: Double?
        let distanceKM: Double?
        let anniversaryYears: Int?
    }

    static func summarize(
        ownerID: UUID,
        from requestedFrom: String,
        to requestedTo: String,
        sessions allSessions: [WorkoutSession],
        logs allLogs: [WorkoutLog],
        importedActivities allImportedActivities: [ImportedActivity]
    ) -> Summary {
        let from = min(requestedFrom, requestedTo)
        let to = max(requestedFrom, requestedTo)
        let allOwnedSessions = allSessions.filter { $0.userID == ownerID && $0.completed }
        let allExternal = allImportedActivities.filter { activity in
            activity.userID == ownerID
                && activity.hiddenAt == nil
                && !ExternalWorkoutImport.isAPEXBundleIdentifier(activity.sourceBundleIdentifier ?? "")
        }
        let sessions = allOwnedSessions.filter { $0.date >= from && $0.date <= to }
        let sessionIDs = Set(sessions.map(\.id))
        let external = allExternal.filter { $0.date >= from && $0.date <= to }
        let linkedSessionIDs = Set(external.compactMap { activity -> UUID? in
            guard let sessionID = activity.apexWorkoutSessionID, sessionIDs.contains(sessionID) else { return nil }
            return sessionID
        })
        let logs = allLogs.filter { log in
            log.userID == ownerID && sessionIDs.contains(log.sessionID) && !log.skipped
        }

        var durationMinutes = external.reduce(0) { total, activity in
            total + max(0, activity.durationMinutes)
        }
        for session in sessions where !linkedSessionIDs.contains(session.id) {
            if let minutes = validDurationMinutes(startedAt: session.startedAt, completedAt: session.completedAt) {
                durationMinutes += minutes
            } else {
                let seconds = logs
                    .filter { $0.sessionID == session.id }
                    .reduce(0) { $0 + max(0, $1.durationSeconds ?? 0) }
                if seconds > 0 { durationMinutes += max(1, Int((Double(seconds) / 60).rounded())) }
            }
        }

        let energyFacts = external.compactMap { activity -> Double? in
            guard let value = activity.activeEnergyKcal, value.isFinite, value >= 0 else { return nil }
            return value
        }
        let externalDistanceFacts = external.compactMap { activity -> Double? in
            guard let value = activity.distanceKM, value.isFinite, value >= 0 else { return nil }
            return value
        }
        let loggedDistanceFacts = logs.compactMap { log -> Double? in
            guard !linkedSessionIDs.contains(log.sessionID),
                  let value = log.distanceMeters, value.isFinite, value >= 0 else { return nil }
            return value
        }
        let reps = logs.reduce(0) { $0 + max(0, $1.reps ?? 0) }
        let volumeFacts = logs.compactMap { log -> Double? in
            guard let weight = log.weightKG, let repetitions = log.reps,
                  weight > 0, repetitions > 0 else { return nil }
            return weight * Double(repetitions)
        }
        let activeDates = Set(sessions.map(\.date) + external.map(\.date))
        let standaloneExternal = external.filter { activity in
            guard let sessionID = activity.apexWorkoutSessionID else { return true }
            return !sessionIDs.contains(sessionID)
        }
        let oldestEvidenceDate = (allOwnedSessions.map(\.date) + allExternal.map(\.date)).min()
        let distanceFacts = externalDistanceFacts.reduce(0, +) + loggedDistanceFacts.reduce(0, +) / 1_000

        return Summary(
            from: from,
            to: to,
            workouts: sessions.count + standaloneExternal.count,
            activeDays: activeDates.count,
            durationMinutes: durationMinutes,
            activeEnergyKcal: energyFacts.isEmpty ? nil : energyFacts.reduce(0, +),
            sets: logs.count,
            reps: reps,
            volumeKG: volumeFacts.isEmpty ? nil : volumeFacts.reduce(0, +),
            distanceKM: externalDistanceFacts.isEmpty && loggedDistanceFacts.isEmpty ? nil : distanceFacts,
            anniversaryYears: anniversaryYears(oldestEvidenceDate: oldestEvidenceDate, from: from, to: to)
        )
    }

    static func anniversaryYears(oldestEvidenceDate: String?, from: String, to: String) -> Int? {
        guard let oldestEvidenceDate,
              let oldest = ISO8601DateFormatter.apexDateOnly.date(from: oldestEvidenceDate),
              let start = ISO8601DateFormatter.apexDateOnly.date(from: from),
              let end = ISO8601DateFormatter.apexDateOnly.date(from: to),
              start <= end else { return nil }
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(secondsFromGMT: 0)!
        for years in [10, 5, 1] {
            guard let boundary = calendar.date(byAdding: .year, value: -years, to: end) else { continue }
            if start <= boundary && oldest <= boundary { return years }
        }
        return nil
    }

    private static func validDurationMinutes(startedAt: String?, completedAt: String?) -> Int? {
        guard let startedAt, let completedAt,
              let start = parseTimestamp(startedAt), let end = parseTimestamp(completedAt), end > start else { return nil }
        return max(1, Int((end.timeIntervalSince(start) / 60).rounded()))
    }

    private static func parseTimestamp(_ value: String) -> Date? {
        let fractional = ISO8601DateFormatter()
        fractional.formatOptions.insert(.withFractionalSeconds)
        return fractional.date(from: value) ?? ISO8601DateFormatter().date(from: value)
    }
}

File: src/lib/workoutInsights.ts
import type { ImportedActivity, WorkoutLog, WorkoutSession } from './types.ts'
import { isAPEXWorkoutSourceBundle } from './wearableWorkoutLinking.ts'

export interface WorkoutInsightInput {
  ownerID: string
  from: string
  to: string
  sessions: WorkoutSession[]
  logs: WorkoutLog[]
  importedActivities: ImportedActivity[]
}

export interface WorkoutInsightSummary {
  from: string
  to: string
  workouts: number
  activeDays: number
  durationMinutes: number
  activeEnergyKcal: number | null
  sets: number
  reps: number
  volumeKg: number | null
  distanceKm: number | null
  anniversaryYears: 1 | 5 | 10 | null
}

function inRange(date: string, from: string, to: string): boolean {
  return date >= from && date <= to
}

function validDurationMinutes(startedAt?: string | null, completedAt?: string | null): number | null {
  if (!startedAt || !completedAt) return null
  const start = Date.parse(startedAt)
  const end = Date.parse(completedAt)
  if (!Number.isFinite(start) || !Number.isFinite(end) || end <= start) return null
  return Math.max(1, Math.round((end - start) / 60_000))
}

function subtractCalendarYears(isoDate: string, years: number): string | null {
  const match = /^(\d{4})-(\d{2})-(\d{2})$/.exec(isoDate)
  if (!match) return null
  const year = Number(match[1]) - years
  const month = Number(match[2])
  const day = Number(match[3])
  const result = new Date(Date.UTC(year, month - 1, day))
  if (result.getUTCMonth() !== month - 1) result.setUTCDate(0)
  return result.toISOString().slice(0, 10)
}

export function anniversaryYearsForWorkoutRange(
  oldestEvidenceDate: string | null,
  from: string,
  to: string,
): 1 | 5 | 10 | null {
  if (!oldestEvidenceDate || from > to) return null
  for (const years of [10, 5, 1] as const) {
    const boundary = subtractCalendarYears(to, years)
    if (boundary && from <= boundary && oldestEvidenceDate <= boundary) return years
  }
  return null
}

function isVisibleExternal(activity: ImportedActivity, ownerID: string): boolean {
  return activity.user_id === ownerID
    && !activity.hidden_at
    && !isAPEXWorkoutSourceBundle(activity.source_bundle_id)
}

export function workoutInsights(input: WorkoutInsightInput): WorkoutInsightSummary {
  const from = input.from <= input.to ? input.from : input.to
  const to = input.from <= input.to ? input.to : input.from
  const allOwnedSessions = input.sessions.filter((session) => session.user_id === input.ownerID && session.completed)
  const allExternal = input.importedActivities.filter((activity) => isVisibleExternal(activity, input.ownerID))
  const sessions = allOwnedSessions.filter((session) => inRange(session.date, from, to))
  const sessionIDs = new Set(sessions.map((session) => session.id))
  const external = allExternal.filter((activity) => inRange(activity.date, from, to))
  const linkedSessionIDs = new Set(external.flatMap((activity) => (
    activity.apex_workout_session_id && sessionIDs.has(activity.apex_workout_session_id)
      ? [activity.apex_workout_session_id]
      : []
  )))
  const logs = input.logs.filter((entry) => (
    entry.user_id === input.ownerID
    && sessionIDs.has(entry.session_id)
    && !entry.skipped
  ))

  let durationMinutes = external.reduce((total, activity) => (
    total + (Number.isFinite(activity.duration_min) && activity.duration_min > 0 ? activity.duration_min : 0)
  ), 0)
  for (const session of sessions) {
    if (linkedSessionIDs.has(session.id)) continue
    const timestampDuration = validDurationMinutes(session.started_at, session.completed_at)
    if (timestampDuration != null) {
      durationMinutes += timestampDuration
      continue
    }
    const loggedSeconds = logs
      .filter((entry) => entry.session_id === session.id)
      .reduce((total, entry) => total + Math.max(0, entry.duration_seconds ?? 0), 0)
    if (loggedSeconds > 0) durationMinutes += Math.max(1, Math.round(loggedSeconds / 60))
  }

  const energyFacts = external
    .map((activity) => activity.active_energy_kcal)
    .filter((value): value is number => value != null && Number.isFinite(value) && value >= 0)
  const externalDistanceFacts = external
    .map((activity) => activity.distance_km)
    .filter((value): value is number => value != null && Number.isFinite(value) && value >= 0)
  const loggedDistanceFacts = logs
    .filter((entry) => !linkedSessionIDs.has(entry.session_id))
    .map((entry) => entry.distance_meters)
    .filter((value): value is number => value != null && Number.isFinite(value) && value >= 0)

  const reps = logs.reduce((total, entry) => total + Math.max(0, entry.reps ?? 0), 0)
  const volumeFacts = logs.flatMap((entry) => {
    const weight = entry.weight_kg
    const repetitions = entry.reps
    return weight != null && repetitions != null && weight > 0 && repetitions > 0
      ? [weight * repetitions]
      : []
  })
  const activeDates = new Set([
    ...sessions.map((session) => session.date),
    ...external.map((activity) => activity.date),
  ])
  const standaloneExternal = external.filter((activity) => (
    !activity.apex_workout_session_id || !sessionIDs.has(activity.apex_workout_session_id)
  ))
  const oldestEvidenceDate = [...allOwnedSessions.map((session) => session.date), ...allExternal.map((activity) => activity.date)]
    .sort()[0] ?? null

  return {
    from,
    to,
    workouts: sessions.length + standaloneExternal.length,
    activeDays: activeDates.size,
    durationMinutes,
    activeEnergyKcal: energyFacts.length > 0 ? energyFacts.reduce((total, value) => total + value, 0) : null,
    sets: logs.length,
    reps,
    volumeKg: volumeFacts.length > 0 ? volumeFacts.reduce((total, value) => total + value, 0) : null,
    distanceKm: externalDistanceFacts.length + loggedDistanceFacts.length > 0
      ? externalDistanceFacts.reduce((total, value) => total + value, 0)
        + loggedDistanceFacts.reduce((total, value) => total + value, 0) / 1_000
      : null,
    anniversaryYears: anniversaryYearsForWorkoutRange(oldestEvidenceDate, from, to),
  }
}
